File: intrinsic/executive/py/behavior_tree_visitor.py
# Copyright 2023 Intrinsic Innovation LLC

"""Visitor implementation for Behavior Trees in Python."""

from collections.abc import Callable

from intrinsic.executive.proto import behavior_tree_pb2

TreeVisitorCallback = Callable[[behavior_tree_pb2.BehaviorTree], None]
NodeVisitorCallback = Callable[
    [behavior_tree_pb2.BehaviorTree, behavior_tree_pb2.BehaviorTree.Node], None
]
ConditionVisitorCallback = Callable[
    [behavior_tree_pb2.BehaviorTree, behavior_tree_pb2.BehaviorTree.Condition],
    None,
]


def _walk_tree(
    tree: behavior_tree_pb2.BehaviorTree,
    tree_visitor: TreeVisitorCallback | None,
    node_visitor: NodeVisitorCallback | None,
    condition_visitor: ConditionVisitorCallback | None,
    visit_called_tree_state: bool,
) -> None:
  """Recursively walks a given tree and invokes visitor.

  Args:
    tree: tree to call visitor for and walk
    tree_visitor: optional callback to invoke for trees
    node_visitor: optional callback to invoke for nodes
    condition_visitor: optional callback to invoke for conditions
    visit_called_tree_state: whether to visit called_tree_state in TaskNodes
  """
  if tree_visitor is not None:
    tree_visitor(tree)

  _walk_node(
      tree,
      tree.root,
      tree_visitor,
      node_visitor,
      condition_visitor,
      visit_called_tree_state,
  )


def _walk_node(
    tree: behavior_tree_pb2.BehaviorTree,
    node: behavior_tree_pb2.BehaviorTree.Node,
    tree_visitor: TreeVisitorCallback | None,
    node_visitor: NodeVisitorCallback | None,
    condition_visitor: ConditionVisitorCallback | None,
    visit_called_tree_state: bool,
) -> None:
  """Recursively walks a given node and invokes visitor.

  Args:
    tree: tree the node belongs to
    node: node to call visitor for and walk
    tree_visitor: optional callback to invoke for trees
    node_visitor: optional callback to invoke for nodes
    condition_visitor: optional callback to invoke for conditions
    visit_called_tree_state: whether to visit called_tree_state in TaskNodes
  """
  if node_visitor is not None:
    node_visitor(tree, node)

  if node.HasField("decorators"):
    if node.decorators.HasField("condition"):
      _walk_condition(
          tree,
          node.decorators.condition,
          tree_visitor,
          node_visitor,
          condition_visitor,
          visit_called_tree_state,
      )

  if node.HasField("sequence"):
    for c in node.sequence.children:
      _walk_node(
          tree,
          c,
          tree_visitor,
          node_visitor,
          condition_visitor,
          visit_called_tree_state,
      )
  elif node.HasField("parallel"):
    for c in node.parallel.children:
      _walk_node(
          tree,
          c,
          tree_visitor,
          node_visitor,
          condition_visitor,
          visit_called_tree_state,
      )
  elif node.HasField("selector"):
    for c in node.selector.children:
      _walk_node(
          tree,
          c,
          tree_visitor,
          node_visitor,
          condition_visitor,
          visit_called_tree_state,
      )
    for c in node.selector.branches:
      if c.HasField("condition"):
        _walk_condition(
            tree,
            c.condition,
            tree_visitor,
            node_visitor,
            condition_visitor,
            visit_called_tree_state,
        )
      _walk_node(
          tree,
          c.node,
          tree_visitor,
          node_visitor,
          condition_visitor,
          visit_called_tree_state,
      )
  elif node.HasField("fallback"):
    for c in node.fallback.children:
      _walk_node(
          tree,
          c,
          tree_visitor,
          node_visitor,
          condition_visitor,
          visit_called_tree_state,
      )
    for c in node.fallback.tries:
      if c.HasField("condition"):
        _walk_condition(
            tree,
            c.condition,
            tree_visitor,
            node_visitor,
            condition_visitor,
            visit_called_tree_state,
        )
      _walk_node(
          tree,
          c.node,
          tree_visitor,
          node_visitor,
          condition_visitor,
          visit_called_tree_state,
      )
  elif node.HasField("branch"):
    if node.branch.HasField("if"):
      _walk_condition(
          tree,
          getattr(node.branch, "if"),
          tree_visitor,
          node_visitor,
          condition_visitor,
          visit_called_tree_state,
      )
    if node.branch.HasField("then"):
      _walk_node(
          tree,
          node.branch.then,
          tree_visitor,
          node_visitor,
          condition_visitor,
          visit_called_tree_state,
      )
    if node.branch.HasField("else"):
      _walk_node(
          tree,
          getattr(node.branch, "else"),
          tree_visitor,
          node_visitor,
          condition_visitor,
          visit_called_tree_state,
      )
  elif node.HasField("loop"):
    if node.loop.HasField("while"):
      _walk_condition(
          tree,
          getattr(node.loop, "while"),
          tree_visitor,
          node_visitor,
          condition_visitor,
          visit_called_tree_state,
      )
    if node.loop.HasField("do"):
      _walk_node(
          tree,
          node.loop.do,
          tree_visitor,
          node_visitor,
          condition_visitor,
          visit_called_tree_state,
      )
  elif node.HasField("retry"):
    if node.retry.HasField("child"):
      _walk_node(
          tree,
          node.retry.child,
          tree_visitor,
          node_visitor,
          condition_visitor,
          visit_called_tree_state,
      )
    if node.retry.HasField("recovery"):
      _walk_node(
          tree,
          node.retry.recovery,
          tree_visitor,
          node_visitor,
          condition_visitor,
          visit_called_tree_state,
      )
  elif node.HasField("sub_tree"):
    if node.sub_tree.HasField("tree"):
      _walk_tree(
          node.sub_tree.tree,
          tree_visitor,
          node_visitor,
          condition_visitor,
          visit_called_tree_state,
      )
  elif node.HasField("task"):
    if visit_called_tree_state and node.task.HasField("called_tree_state"):
      _walk_tree(
          node.task.called_tree_state,
          tree_visitor,
          node_visitor,
          condition_visitor,
          visit_called_tree_state,
      )


def _walk_condition(
    tree: behavior_tree_pb2.BehaviorTree,
    cond: behavior_tree_pb2.BehaviorTree.Condition,
    tree_visitor: TreeVisitorCallback | None,
    node_visitor: NodeVisitorCallback | None,
    condition_visitor: ConditionVisitorCallback | None,
    visit_called_tree_state: bool,
) -> None:
  """Recursively walks a given condition and invokes visitor.

  Args:
    tree: tree the condition belongs to
    cond: condition to call visitor for and walk
    tree_visitor: optional callback to invoke for trees
    node_visitor: optional callback to invoke for nodes
    condition_visitor: optional callback to invoke for conditions
    visit_called_tree_state: whether to visit called_tree_state in TaskNodes
  """
  if condition_visitor is not None:
    condition_visitor(tree, cond)

  if cond.HasField("behavior_tree"):
    _walk_tree(
        cond.behavior_tree,
        tree_visitor,
        node_visitor,
        condition_visitor,
        visit_called_tree_state,
    )
  elif cond.HasField("all_of"):
    for c in cond.all_of.conditions:
      _walk_condition(
          tree,
          c,
          tree_visitor,
          node_visitor,
          condition_visitor,
          visit_called_tree_state,
      )
  elif cond.HasField("any_of"):
    for c in cond.any_of.conditions:
      _walk_condition(
          tree,
          c,
          tree_visitor,
          node_visitor,
          condition_visitor,
          visit_called_tree_state,
      )
  elif cond.HasField("not"):
    _walk_condition(
        tree,
        getattr(cond, "not"),
        tree_visitor,
        node_visitor,
        condition_visitor,
        visit_called_tree_state,
    )


def walk(
    tree: behavior_tree_pb2.BehaviorTree,
    *,
    tree_visitor: TreeVisitorCallback | None = None,
    node_visitor: NodeVisitorCallback | None = None,
    condition_visitor: ConditionVisitorCallback | None = None,
    visit_called_tree_state=False,
) -> None:
  """Recursively walks a given tree and invokes visitors.

  Args:
    tree: tree to walk
    tree_visitor: optional callback to invoke for trees
    node_visitor: optional callback to invoke for nodes
    condition_visitor: optional callback to invoke for conditions
    visit_called_tree_state: whether to visit called_tree_state in TaskNodes
  """
  _walk_tree(
      tree,
      tree_visitor,
      node_visitor,
      condition_visitor,
      visit_called_tree_state,
  )

File: intrinsic/executive/go/behavior_tree_visitor.go
// Copyright 2023 Intrinsic Innovation LLC

// Package behaviortree provides utilities to operate on Behavior Trees.
//
// Features are:
// - Enables to walk and execute code for each node and condition in the tree.
package behaviortree

import (
	"context"
	"errors"
	"iter"

	"google.golang.org/protobuf/proto"

	btpb "intrinsic/executive/proto/behavior_tree_go_proto"
)

type visitElement struct {
	msg      proto.Message
	ancestor *VisitElement
}

// VisitElement is a single element visited by the [Visitor] when using [Walk].
//
// One of:
// - [btpb.BehaviorTree]
// - [btpb.BehaviorTree_Node]
// - [btpb.BehaviorTree_Condition]
type VisitElement visitElement

// VisitElementFromTree wraps a tree into a [VisitElement].
func VisitElementFromTree(tree *btpb.BehaviorTree) VisitElement {
	return VisitElement{tree, nil}
}

// VisitElementFromNode wraps a node into a [VisitElement].
func VisitElementFromNode(node *btpb.BehaviorTree_Node) VisitElement {
	return VisitElement{node, nil}
}

// VisitElementFromCondition wraps a condition into a [VisitElement].
func VisitElementFromCondition(cond *btpb.BehaviorTree_Condition) VisitElement {
	return VisitElement{cond, nil}
}

// AsAncestorForTree returns a new visit element for the given tree with this
// element as its ancestor. Panics if the visit element is not valid (i.e.
// [VisitElement.IsValid] would return `false`).
func (v VisitElement) AsAncestorForTree(tree *btpb.BehaviorTree) VisitElement {
	if !v.IsValid() {
		panic("cannot use invalid visit element as ancestor for tree")
	}
	return VisitElement{tree, &v}
}

// AsAncestorForNode returns a new visit element for the given node with this
// element as its ancestor. Panics if the visit element is not valid (i.e.
// [VisitElement.IsValid] would return `false`).
func (v VisitElement) AsAncestorForNode(node *btpb.BehaviorTree_Node) VisitElement {
	if !v.IsValid() {
		panic("cannot use invalid visit element as ancestor for node")
	}
	return VisitElement{node, &v}
}

// AsAncestorForCondition returns a new visit element for the given condition
// with this element as its ancestor. Panics if the visit element is not valid
// (i.e. [VisitElement.IsValid] would return `false`).
func (v VisitElement) AsAncestorForCondition(cond *btpb.BehaviorTree_Condition) VisitElement {
	if !v.IsValid() {
		panic("cannot use invalid visit element as ancestor for condition")
	}
	return VisitElement{cond, &v}
}

// IsValid returns whether the element is valid. The element is considered valid
// if it holds a tree, node or condition.
func (v VisitElement) IsValid() bool {
	return v.Tree() != nil || v.Node() != nil || v.Condition() != nil
}

// Tree returns the tree if the element represents one or nil otherwise.
func (v VisitElement) Tree() *btpb.BehaviorTree {
	if tree, ok := v.msg.(*btpb.BehaviorTree); ok {
		return tree
	}
	return nil
}

// Node returns the node if the element represents one or nil otherwise.
func (v VisitElement) Node() *btpb.BehaviorTree_Node {
	if node, ok := v.msg.(*btpb.BehaviorTree_Node); ok {
		return node
	}
	return nil
}

// Condition returns the condition if the element represents one or nil
// otherwise.
func (v VisitElement) Condition() *btpb.BehaviorTree_Condition {
	if cond, ok := v.msg.(*btpb.BehaviorTree_Condition); ok {
		return cond
	}
	return nil
}

// Ancestor returns the direct ancestor of the element. Returns `nil` if the
// element does not have any ancestor.
func (v VisitElement) Ancestor() *VisitElement {
	return v.ancestor
}

// Ancestors returns an iterator over the ancestors of the element. Yields items
// in order from the direct anecestor of the element up to the root element.
// Does not yield elements that are not valid (i.e. [VisitElement.IsValid]
// returns `false`).
func (v VisitElement) Ancestors() iter.Seq[VisitElement] {
	return func(yield func(VisitElement) bool) {
		ancestor := v.ancestor
		for ancestor != nil {
			if !ancestor.IsValid() {
				return
			}
			if !yield(*ancestor) {
				return
			}
			ancestor = ancestor.ancestor
		}
	}
}

// Ancestors returns an iterator over the ancestors of the element. Yields items
// in order from the direct anecestor of the element up to the root element.
// Does not yield elements that are not valid (i.e. [VisitElement.IsValid]
// returns `false`).
func (v VisitElement) AncestorsFromRoot() iter.Seq[VisitElement] {
	return func(yield func(VisitElement) bool) {
		ancestor := v.ancestor
		for ancestor != nil {
			if !ancestor.IsValid() {
				return
			}
			if !yield(*ancestor) {
				return
			}
			ancestor = ancestor.ancestor
		}
	}
}

// Stop causes visiting to stop immediately. [Walk] will not return this error.
var Stop = errors.New("stop visiting")

// Visitor defines requirements for visitor implementations for the BehaviorTree
// proto walker. Use it with [Walk] to visit all nodes in a behavior tree.
type Visitor interface {
	// Visit each element in the tree. If this returns any non-nil error (other
	// than [Stop]), it stops visiting and the error is returned by [Walk].
	Visit(ctx context.Context, element VisitElement) error
}

type options struct {
	visitCalledTreeState bool
}

// Option enables customizing the behavior of [Walk].
type Option func(*options)

// VisitCalledTreeState is an option that makes [Walk] traverse into tree in the
// `called_tree_state` field of task nodes when present.
func VisitCalledTreeState() Option {
	return func(o *options) {
		o.visitCalledTreeState = true
	}
}

func walkCondition(ctx context.Context, ancestor VisitElement, cond *btpb.BehaviorTree_Condition, visitor Visitor, opts *options) error {
	if cond == nil {
		return nil
	}
	if err := ctx.Err(); err != nil {
		return err
	}
	element := ancestor.AsAncestorForCondition(cond)
	if err := visitor.Visit(ctx, element); err != nil {
		return err
	}
	switch cond.ConditionType.(type) {
	case *btpb.BehaviorTree_Condition_BehaviorTree:
		err := walkTree(ctx, element, cond.GetBehaviorTree(), visitor, opts)
		if err != nil {
			return err
		}

	case *btpb.BehaviorTree_Condition_AllOf:
		for _, c := range cond.GetAllOf().GetConditions() {
			if err := walkCondition(ctx, element, c, visitor, opts); err != nil {
				return err
			}
		}

	case *btpb.BehaviorTree_Condition_AnyOf:
		for _, c := range cond.GetAnyOf().GetConditions() {
			if err := walkCondition(ctx, element, c, visitor, opts); err != nil {
				return err
			}
		}

	case *btpb.BehaviorTree_Condition_Not:
		if err := walkCondition(ctx, element, cond.GetNot(), visitor, opts); err != nil {
			return err
		}
	}

	return nil
}

func walkNode(ctx context.Context, ancestor VisitElement, node *btpb.BehaviorTree_Node, visitor Visitor, opts *options) error {
	if node == nil {
		return nil
	}
	if err := ctx.Err(); err != nil {
		return err
	}
	element := ancestor.AsAncestorForNode(node)
	if err := visitor.Visit(ctx, element); err != nil {
		return err
	}

	if node.GetDecorators().GetCondition() != nil {
		if err := walkCondition(ctx, element, node.GetDecorators().GetCondition(), visitor, opts); err != nil {
			return err
		}
	}

	switch node.NodeType.(type) {
	case *btpb.BehaviorTree_Node_Sequence:
		for _, c := range node.GetSequence().GetChildren() {
			if err := walkNode(ctx, element, c, visitor, opts); err != nil {
				return err
			}
		}

	case *btpb.BehaviorTree_Node_Parallel:
		for _, c := range node.GetParallel().GetChildren() {
			if err := walkNode(ctx, element, c, visitor, opts); err != nil {
				return err
			}
		}

	case *btpb.BehaviorTree_Node_Selector:
		for _, c := range node.GetSelector().GetChildren() {
			if err := walkNode(ctx, element, c, visitor, opts); err != nil {
				return err
			}
		}
		for _, c := range node.GetSelector().GetBranches() {
			if err := walkCondition(ctx, element, c.GetCondition(), visitor, opts); err != nil {
				return err
			}
			if err := walkNode(ctx, element, c.GetNode(), visitor, opts); err != nil {
				return err
			}
		}

	case *btpb.BehaviorTree_Node_Fallback:
		for _, c := range node.GetFallback().GetChildren() {
			if err := walkNode(ctx, element, c, visitor, opts); err != nil {
				return err
			}
		}
		for _, c := range node.GetFallback().GetTries() {
			if err := walkCondition(ctx, element, c.GetCondition(), visitor, opts); err != nil {
				return err
			}
			if err := walkNode(ctx, element, c.GetNode(), visitor, opts); err != nil {
				return err
			}
		}

	case *btpb.BehaviorTree_Node_Branch:
		if err := walkCondition(ctx, element, node.GetBranch().GetIf(), visitor, opts); err != nil {
			return err
		}
		if err := walkNode(ctx, element, node.GetBranch().GetThen(), visitor, opts); err != nil {
			return err
		}
		if err := walkNode(ctx, element, node.GetBranch().GetElse(), visitor, opts); err != nil {
			return err
		}

	case *btpb.BehaviorTree_Node_Loop:
		if err := walkCondition(ctx, element, node.GetLoop().GetWhile(), visitor, opts); err != nil {
			return err
		}
		if err := walkNode(ctx, element, node.GetLoop().GetDo(), visitor, opts); err != nil {
			return err
		}

	case *btpb.BehaviorTree_Node_Retry:
		if err := walkNode(ctx, element, node.GetRetry().GetChild(), visitor, opts); err != nil {
			return err
		}
		if err := walkNode(ctx, element, node.GetRetry().GetRecovery(), visitor, opts); err != nil {
			return err
		}

	case *btpb.BehaviorTree_Node_SubTree:
		if err := walkTree(ctx, element, node.GetSubTree().GetTree(), visitor, opts); err != nil {
			return err
		}

	case *btpb.BehaviorTree_Node_Task:
		if opts.visitCalledTreeState {
			if called := node.GetTask().GetCalledTreeState(); called != nil {
				// Trees in called tree state are relevant for finding a node by tree ID.
				// They are therefore considered a new enclosing tree.
				if err := walkTree(ctx, element, called, visitor, opts); err != nil {
					return err
				}
			}
		}
	}

	return nil
}

func walkTree(ctx context.Context, ancestor VisitElement, tree *btpb.BehaviorTree, visitor Visitor, opts *options) error {
	var element VisitElement
	if ancestor.IsValid() {
		element = ancestor.AsAncestorForTree(tree)
	} else {
		element = VisitElementFromTree(tree)
	}
	if err := visitor.Visit(ctx, element); err != nil {
		return err
	}
	return walkNode(ctx, element, tree.Root, visitor, opts)
}

// Walk walks the given Behavior Tree and invokes the given visitor for nodes
// and conditions of the tree.
func Walk(ctx context.Context, tree *btpb.BehaviorTree, visitor Visitor, opts ...Option) error {
	options := &options{}
	for _, opt := range opts {
		opt(options)
	}
	err := walkTree(ctx, VisitElement{}, tree, visitor, options)
	if err == Stop {
		err = nil
	}
	return err
}
